### PR TITLE
Optimize the web demo for yolov4

### PR DESCRIPTION
**Pull request by dvartaniansTT**
### Problem description
Have a real-time web demo for yolov4

### What's changed
Enable trace + 2cq
Optimize the post processing 

### Checklist
- [x] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

**Comment by tt-rkim** [comment]
Please run post commit and nightly fast dispatch
You also seem to be on quite an old base of main.

**Comment by dvartaniansTT** [comment]
@tt-rkim I have rebased to main now and post-commits and nightly fast dispatch are currently running. 
https://github.com/tenstorrent/tt-metal/actions/runs/12043538909
https://github.com/tenstorrent/tt-metal/actions/runs/12043531296

nothing failing yet. will update you if anything fails.

**Review by mbahnasTT — APPROVED** [review]

**Comment by dvartaniansTT** [comment]
@tt-rkim , I have resolved the CI test issues relevant to yolo-v4. and launched the CI tests again: 

[nightly fast dispatch tests](https://github.com/tenstorrent/tt-metal/actions/runs/12129527247/job/33818961909)
[all post commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/12130429711)

File: models/demos/yolov4/demo/demo.py [hunk 1]
```diff
@@ -550,13 +550,15 @@ def do_detect(model, img, conf_thresh, nms_thresh, n_classes, device=None, class
             print("-----------------------------------")
 
             boxes = post_processing(img, conf_thresh, nms_thresh, output)
+            print(boxes)
```
Comment (TT-BrianLiu) [inline models/demos/yolov4/demo/demo.py:553]: This is intentional change?

File: models/demos/yolov4/ttnn/genboxes.py [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+import torch
+import math
+import numpy as np
+import ttnn
+
+
+def _nearest_32(x):
+    return math.ceil(x / 32) * 32
+
```
Comment (TT-BrianLiu) [inline models/demos/yolov4/ttnn/genboxes.py:13]: This should exist somewhere. Try to not duplicate code

File: models/demos/yolov4/web_demo/server/fast_api_yolov4.py [hunk 3]
```diff
@@ -60,11 +180,23 @@ async def objdetection_v2(file: UploadFile = File(...)):
...
+    output = boxes[0]
+    # output = boxes
+    try:
+        output = process_output(output)
+    except Exception as E:
+        print("the Exception is: ", E)
+        print("No objects detected!")
+        return []
+    t3 = time.time()
+    print("the post processing to get the boexes took: ", t3 - t2)
```
Comment (TT-BrianLiu) [inline models/demos/yolov4/web_demo/server/fast_api_yolov4.py:201]: Logger?

File: models/demos/yolov4/web_demo/client/yolov4.py [hunk 4]
```diff
@@ -196,52 +94,61 @@ def get_color(c, x, max_val):
...
+                    output = orjson.loads(response.content)
+                else:
+                    print(f"Request failed with status code {response.status_code}")
+                    # return None
+        except requests.exceptions.RequestException as e:
+            print(f"Request failed: {e}")
+            return None
 
         t3 = time.time()
+        print("Request and parsing took: ", t3 - t10)
```
Comment (TT-BrianLiu) [inline models/demos/yolov4/web_demo/client/yolov4.py:137]: Logger?

File: tests/ttnn/integration_tests/yolov4/test_ttnn_yolov4.py [hunk 5]
```diff
@@ -47,27 +98,74 @@ def test_yolov4(device, reset_seeds, model_location_generator):
...
+
+    result_boxes_padded = result_boxes_padded.permute(0, 2, 1, 3)
+    result_boxes_list = []
+    result_boxes_list.append(result_boxes_padded[:, 0:6100])
+    result_boxes_list.append(result_boxes_padded[:, 6128:6228])
+    result_boxes_list.append(result_boxes_padded[:, 6256:6356])
+    result_boxes = torch.cat(result_boxes_list, dim=1)
+
+    print(ref_boxes.shape, ref_confs.shape)
+    print(result_boxes.shape, result_confs.shape)
```
Comment (TT-BrianLiu) [inline tests/ttnn/integration_tests/yolov4/test_ttnn_yolov4.py:118]: Remove or add clearer message and switch to logger?

**Review by TT-BrianLiu — APPROVED** [review]
Minor comments

**Comment by tt-rkim** [comment]
Let's see what's happening with the re-run of nightly fast dispatch: https://github.com/tenstorrent/tt-metal/actions/runs/12129527247
It's taking a long time for GS ttnn integration tests - very suspect.

**Review by tt-rkim — APPROVED** [review]

**Review by bgoelTT — CHANGES_REQUESTED** [review]
`orjson` needs to be added to the client's dependencies, I'm not sure what version you used for testing @dvartaniansTT but I got `orjson==3.10.12` to work.

File: models/demos/yolov4/web_demo/server/fast_api_yolov4.py [hunk 6]
```diff
@@ -25,31 +27,156 @@ async def root():
...
+    if ("WH_ARCH_YAML" in os.environ) and os.environ["WH_ARCH_YAML"] == "wormhole_b0_80_arch_eth_dispatch.yaml":
+        print("WH_ARCH_YAML:", os.environ.get("WH_ARCH_YAML"))
+        device_id = 0
+        device = ttnn.CreateDevice(
+            device_id,
+            dispatch_core_type=get_dispatch_core_type(),
+            l1_small_size=24576,
+            trace_region_size=3205120,
+            num_command_queues=2,
+        )
```
Comment (bgoelTT) [inline models/demos/yolov4/web_demo/server/fast_api_yolov4.py:57]: I think the `ttnn.CreateDevice()` API has changed. It not longer take the `dispatch_core_type` kwarg. Instead it now takes a `dispatch_core_config` kwarg which expects an object of type `ttnn.DispatchCoreConfig`. Grayskull and Wormhole devices use `dispatch_core_axis=ttnn.DispatchCoreAxis.ROW` whereas Blackhole will use `dispatch_core_axis=ttnn.DispatchCoreAxis.COL` as per `conftest.py`. I got things working locally, see my suggestion below:
```suggestion
        device_id = 0
        dispatch_core_config = ttnn.DispatchCoreConfig(get_dispatch_core_type(), ttnn.DispatchCoreAxis.ROW)
        device = ttnn.CreateDevice(
            device_id,
            dispatch_core_config=dispatch_core_config,
            l1_small_size=24576,
            trace_region_size=3205120,
            num_command_queues=2,
        )
```

**Review by bgoelTT — CHANGES_REQUESTED** [review]

**Comment by aliuTT** [comment]
As a reminder, `WH_ARCH_YAML` is a deprecated env flag for Metal. The plan is to remove this flag from all codebase once this ticket is resolved https://github.com/tenstorrent/tt-metal/issues/11059. Currently, blocked by models team to remove all use cases of this flag at the python testing level.

File: models/demos/yolov4/web_demo/server/fast_api_yolov4.py (hunk 6 above)
Comment (dvartaniansTT) [reply models/demos/yolov4/web_demo/server/fast_api_yolov4.py:57]: thanks @bgoelTT , yes we noticed this change as well yesterday. we will update the PR.

**Comment by dvartaniansTT** [comment]
Thanks for the headsup @aliuTT , what is the alternative way to check if we are on a N150 or N300? and how to create/open device in each case?

**Review by bgoelTT — APPROVED** [review]

**Comment by dvartaniansTT** [comment]
[all post commits CI pass](https://github.com/tenstorrent/tt-metal/actions/runs/12132373816)

**Review by Denys88 — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

**Review by cfjchu — APPROVED** [review]

**Review by uaydonat — APPROVED** [review]

**Pull request merged**